overlord/snapstate/snapmgr.go

```diff
@@ -99,9 +104,46 @@ func (m *SnapManager) doInstallSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 	t.State().Unlock()
 
+	// see if it is a local snap, if we, nothing to download
+	if fi, err := os.Stat(inst.Name); err == nil && fi.Mode().IsRegular() {
+		inst.DownloadedSnap = inst.Name
+		inst.Developer = snappy.SideloadedDeveloper
+	} else {
+		pb := &TaskProgressAdapter{task: t}
+		downloadedSnapFile, developer, err := m.backend.Download(inst.Name, inst.Channel, pb)
+		if err != nil {
+			return err
+		}
+		inst.DownloadedSnap = downloadedSnapFile
+		inst.Developer = developer
+	}
+
+	// update instState for the next task
+	t.State().Lock()
+	t.Set("install-state", inst)
```

Comment:
"install-state" here feels a bit like overspecializing the task too early. Downloading may end up being done on itself for other reasons, and we may also end up having to diverge the download and the install state structs for handling particular needs of each task more appropriately.

I'd keep using installState-the-struct for the time being, since that works for now, but would rename the state field to "download-state".